### ISSUE TITLE
More complete example please?

### ISSUE THREAD

**Issue by hughperkins** [facebookresearch/higher#108]
Hi, dont suppose.... could it be possible to have a more complete, yet still very simple, example of how to use higher please?
- ~like how would we write MAML using higher? (I'm aware that some of the linked papers use MAML, but the implementation tends to be buried in a lot of paper-specific things, e.g. https://github.com/facebookresearch/LearningToLearn/blob/main/ml3/ml3_train.py#L17-L34 has a bunch of abstract functions etc, without descriptions provided of what they are, so one has to go hunting through the code to figure out what they do, etc. Similarly for https://github.com/Nithin-Holla/MetaWSD/blob/master/models/seq_meta.py#L120-L143~  Ooohh, there is a maml example at https://github.com/facebookresearch/higher/blob/master/examples/maml-omniglot.py, and I've created a PR to add a link to this in the readme https://github.com/facebookresearch/higher/pull/109
- if one has a loss function that takes as input parameters onto which one backpropped a few times, how would such an example look, in the general case?
    - like eg lets say one had neuralstyle, https://github.com/jcjohnson/neural-style , and one wanted to create a loss function based on the generated image, how would one do that using `higher`?

PS super awesome repo :)

**Comment by hughperkins** [comment]
I suppose one thing that is unclear in my mind is, what is the difference between `diffopt.step(loss)`, and `loss.backward()`? Like, what concretely happens under the hood? Looking at the MAML example, it looks like the forward prop, and the loss function are basically *identical*, and the only difference between the inner loop and the outer meta loop is taht one calls `diffopt.step(loss)`, and one calls `loss.backward()`? what happens concretely inside these two method calls? how do they differ? i.e. here https://github.com/facebookresearch/higher/blob/master/examples/maml-omniglot.py#L147-L165

```python
meta_opt.zero_grad():
for i in range(task_num):
    with higher.innerloop_ctx(
        net, inner_opt, copy_initial_weights=False
    ) as (fnet, diffopt):
        for _ in range(n_inner_iter):
            spt_logits = fnet(x_spt[i])
            spt_loss = F.cross_entropy(spt_logits, y_spt[i])
            diffopt.step(spt_loss)

        qry_logits = fnet(x_qry[i])
        qry_loss = F.cross_entropy(qry_logits, y_qry[i])
        qry_loss.backward()
meta_opt.step()
```

Edit: I guess maybe it is something like, `fnet` has a *copy* of the parameters of the original net? `diffopt.step` updates these copied parameters, after first calculating the gradient of the loss wrt these copied parameters? `loss.backward` calculates the gradient of the loss w.r.t the original parameters?

**Comment by egrefen** [comment]
Hello Hugh. Thanks for your pull request helping address your first question. Regarding your second question, could you give more detail? I'm not sure I understand it.

**Comment by egrefen** [comment]
> I suppose one thing that is unclear in my mind is, what is the difference between `diffopt.step(loss)`, and `loss.backward()`? 

`loss.backward()` is just pytorch's normal backwards function, i.e. it will propagate the gradient of the `loss` backwards to the graph to obtain the gradient with regard to any leaf node (variables with `requires_grad=True`), and store it in that leaf node's `.grad` attribute.

In contrast, `diffopt.step(loss)` does three things:
1. It computes the gradients of `loss` with regard to a set of tensors which the differentiable optimizer treats as being model parameters (these are typically not leaf nodes).
2. It "applies" these gradients to those parameters by creating a new set of "parameters" (again, not leaf nodes) with `grad_fn`s linking them back  to the "older parameters".
3. It updates the pointer to the "latest" version of the paramters in the `fnet` it holds a view over, allowing you to write the rest of your code in canonical pytorch.

I hope this clarifies things.

**Comment by hughperkins** [comment]
Thank you for the explanations. I think I've pieced everything together given these. Super awesome repo. Thank you :)

**Comment by GuodongFan** [comment]
for _ in range(n_inner_iter):
            spt_logits = fnet(x_spt[i])
            spt_loss = F.cross_entropy(spt_logits, y_spt[i])
            diffopt.step(spt_loss)

I do not understand this step.
why train the fnet model using the same data (x_spt[i]) for n_inner_iter times?

**Comment by GuodongFan** [comment]
@hughperkins @egrefen

**Comment by hughperkins** [comment]
You're thinking, "if its the same data, the  gradient will always be in the same direction, so I should just increase the learning rate and do it once", right?

This works if the loss surface is like a bowl, with rotational symmetry.

However, imagine squashing the bowl, a lot, so now you have something more like a long trench, or groove. Now, the initial gradient will point towards the middle of the trench. But you still have to work your way along the bottom of the trench, to reach the minimum.

**Comment by hughperkins** [comment]
(and so therefore you need multiple gradient steps)

**Comment by hughperkins** [comment]
(like something more like this shape, https://www.patioliving.com/jamie-young-company-marble-decorative-plates-and-bowls-accessorie-jyc7longbowh  , and you're near the end, on one side)

<img width="618" alt="Screen Shot 2023-02-11 at 03 50 12" src="https://user-images.githubusercontent.com/123560/218249326-35c3f8d8-2ce3-4498-93e3-727b299cb522.png">

**Comment by GuodongFan** [comment]
Thanks! @hughperkins
And, because this model is not used in the end,  we don't worry about overfitting?

**Comment by hughperkins** [comment]
The target task is to take an unseen training dataset, run `n_inner_iter` steps of gradient descent, then test performance on a test dataset. The input to this task, other than the datasets, is a set of parameters that we learn.

The meta-training task learns this set of parameters, but I don't think it learns a learning rate or similar?

Therefore I feel this could totally overfit, sure. The exact choice of number of iterations, learning rate etc is an emprical question I think?

I suppose it could be interesting to explore throwing in things like learning rate into the set of meta-learned parameters? (number of iterations would be more challenging to learn, since it's discrete => seems likely to need either RL or Gumbel?).

**Comment by hughperkins** [comment]
(note that I haven't kept up with meta-learning literature in years, so it's totally possible someone has already explored these explorations)

**Comment by GuodongFan** [comment]
Ok, I get it, thank you!
